# Pipeline de extracción de ratios financieros — PDF

Notebook correspondiente al **experimento de extracción desde PDF** descrito en la sección 4.4 de la memoria del TFM.

El objetivo es extraer las 22 partidas contables y calcular los mismos ratios que en `excel_pipeline.ipynb` e `imagen_pipeline.ipynb`, cuando el estado financiero llega en formato PDF.

## Empresa de prueba
**Viscofan, S.A.** — PDF de cuentas anuales individuales 2024 descargado directamente de la CNMV. Se eligió este documento por su dificultad: el balance aparece partido entre páginas y las tablas de la PyG tienen líneas superpuestas sobre los valores, generando ruido OCR apreciable.

## Por qué el PDF es el caso más exigente

A diferencia de la imagen (donde toda la información está en un único plano visual), un PDF puede tener:
- **Tablas cortadas entre páginas**: los primeros epígrafes de un bloque al pie de una página, los últimos en el encabezado de la siguiente.
- **Calidad de exportación variable**: líneas de tabla superpuestas con valores, tipografías compactas que el OCR concatena sin espacios.
- **Docling en modo nativo falla**: cuando las tablas se cortan entre páginas, Docling no asocia los nombres de las partidas con sus valores y en ocasiones los omite por completo.

## Estrategia: PDF → imágenes → pipeline de imagen

```
PDF
 │
 ├─ Enfoque 1: Docling directo sobre PDF       ← DESCARTADO
 │                                                (tablas sin nombres de partidas)
 │
 └─ Enfoque 2 (adoptado): página a página
        │
        ├─ Cada página → PNG                   ← PIL / pdf2image
        │
        ├─ Extracción estructurada por página  ← Docling (sin recortes adicionales)
        │
        ├─ Detección de layout                 ← LLM textual
        │
        ├─ DataFrame de conceptos              ← Python determinista
        │
        ├─ Selección de 22 partidas            ← LLM con índice posicional
        │
        ├─ Extracción de valores               ← Python determinista
        │
        └─ Cálculo de 22 ratios               ← Python determinista
```

> **Por qué no hacen falta recortes:** cada página del PDF tiene una densidad de filas mucho menor que una imagen que comprime todo el balance en un único plano. Con esa densidad, Docling identifica las filas correctamente sin necesidad de segmentación adicional (ver `experimentos_recortes.ipynb`).

## Nota sobre la amortización
En la PyG de Viscofan la amortización aparece con **signo negativo** (es un gasto). El OCR la extrae así, por lo que la fórmula aplica `abs(amort)` para que `EBITDA = EBIT - abs(amort)` reste correctamente. Sin este ajuste el EBITDA quedaría inflado en el doble de la amortización.

---


---

## Enfoque 1 (descartado) — Docling directamente sobre el PDF

Primera aproximación: aprovechar el soporte nativo de Docling para PDFs y aplicar `DocumentConverter` directamente sobre el fichero.

**Resultado:** Docling detecta 5 tablas en el documento, pero la tabla del patrimonio neto y pasivo llega sin los nombres de las partidas — solo los valores numéricos. Esto ocurre porque el nombre de cada partida está muy separado visualmente del valor correspondiente y la tabla se corta entre páginas, haciendo que Docling no los asocie o directamente los omita.

Sin los nombres de las partidas es imposible saber qué valor corresponde a qué epígrafe contable, así que este enfoque no es válido para el pipeline.


In [4]:
import pandas as pd
import json
import os
from docling.document_converter import DocumentConverter

# ── CONFIGURACIÓN ─────────────────────────────────────────────────────────────
RUTA_PDF = r"C:\Users\perdi\Desktop\tfm\TFM PYTHON\outputs\balance_2024.pdf"
# ──────────────────────────────────────────────────────────────────────────────

converter = DocumentConverter()
result = converter.convert(RUTA_PDF)

print(f"📄 {os.path.basename(RUTA_PDF)}")
print(f"  Tablas detectadas: {len(result.document.tables)}")

dfs_trozos = []
for i, tabla in enumerate(result.document.tables, 1):
    print(f"\n── Tabla {i} ──")
    df_trozo = tabla.export_to_dataframe()
    df_trozo = pd.DataFrame(
        [df_trozo.columns.tolist()] + df_trozo.values.tolist()
    )
    df_trozo.columns = [f"C{j}" for j in range(df_trozo.shape[1])]
    display(df_trozo)
    dfs_trozos.append(df_trozo)

print(f"\n✔ Total dataframes extraídos: {len(dfs_trozos)}")

[INFO] 2026-05-06 10:04:46,666 [RapidOCR] base.py:22: Using engine_name: onnxruntime
[INFO] 2026-05-06 10:04:46,671 [RapidOCR] download_file.py:60: File exists and is valid: C:\Users\perdi\Desktop\tfm\TFM PYTHON\entornotfm\Lib\site-packages\rapidocr\models\ch_PP-OCRv4_det_mobile.onnx
[INFO] 2026-05-06 10:04:46,671 [RapidOCR] main.py:57: Using C:\Users\perdi\Desktop\tfm\TFM PYTHON\entornotfm\Lib\site-packages\rapidocr\models\ch_PP-OCRv4_det_mobile.onnx
[INFO] 2026-05-06 10:04:46,775 [RapidOCR] base.py:22: Using engine_name: onnxruntime
[INFO] 2026-05-06 10:04:46,777 [RapidOCR] download_file.py:60: File exists and is valid: C:\Users\perdi\Desktop\tfm\TFM PYTHON\entornotfm\Lib\site-packages\rapidocr\models\ch_ppocr_mobile_v2.0_cls_mobile.onnx
[INFO] 2026-05-06 10:04:46,777 [RapidOCR] main.py:57: Using C:\Users\perdi\Desktop\tfm\TFM PYTHON\entornotfm\Lib\site-packages\rapidocr\models\ch_ppocr_mobile_v2.0_cls_mobile.onnx
[INFO] 2026-05-06 10:04:46,822 [RapidOCR] base.py:22: Using engine_nam

📄 balance_2024.pdf
  Tablas detectadas: 5

── Tabla 1 ──


,C0,C1,C2,C3
0,ACTIVO,Notas,Periodo.2024,Periodo.2023
1,Inmovilizado Intangible,5,8.846,8.862
2,Aplicaciones informáticas,,8.769,8.862
3,Anticipos y activos en curso,,77,0
4,Inmovilizado Material,,924,1.246
5,Instalaciones técnicas y otro inmovilizado mat...,,923,1.246
6,Anticipos y activos en curso,,1,0
7,Inversiones financieras en empresas del grupo ...,,560.000,548.028
8,Instrumentos de patrimonio,6,560.000,547.236
9,Créditos a empresas,7,0,792


Usage of TableItem.export_to_dataframe() without `doc` argument is deprecated.



── Tabla 2 ──


,C0,C1,C2
0,0,1,2
1,9.1,32.550,32.550
2,,32.550,32.550
3,9.2,12,12
4,9.3,500.160,523.464
5,9.5,-35.045,-21.671
6,3,107.472,151.362
7,3.1,-26.844,-64.563
8,9.6,3.696,2.358
9,,582.004,623.512


Usage of TableItem.export_to_dataframe() without `doc` argument is deprecated.



── Tabla 3 ──


,C0,C1,C2,C3
0,(Miles de euros).,Notas,Periodo.2.024,Periodo.2.023
1,OPERACIONES CONTINUADAS,,,
2,Importe neto de la cifra de negocios,13.1,153.786,191.750
3,Ingresos de participaciones en instrumentos de...,,122.363,161.853
4,Prestación de servicios y otros ingresos,,31.423,29.897
5,Aprovisionamientos,,-440,-184
6,Consumo de mercaderías,13.2,-440,-184
7,Otros ingresos de explotación,,22,31
8,Subvenciones de explotación incorporadas al re...,,22,31
9,Gastos de personal,13.3,-23.901,-21.529


Usage of TableItem.export_to_dataframe() without `doc` argument is deprecated.



── Tabla 4 ──


,C0,C1,C2,C3
0,RESULTADO ANTES DE IMPUESTOS,,111.472,155.697
1,Gasto por impuesto sobre las ganancias,12,-4.000,-4.335
2,RESULTADO DEL EJERCICIO PROCEDENTE DE OPERACIO...,,107.472,151.362
3,RESULTADO DEL EJERCICIO,,107.472,151.362


Usage of TableItem.export_to_dataframe() without `doc` argument is deprecated.



── Tabla 5 ──


,C0,C1,C2,C3
0,(Miles de euros),Notas,Periodo.2024,Periodo.2023
1,RESULTADO DE LA CUENTA DE PERDIDAS Y GANANCIAS,,107.472,151.362
2,INGRESOS Y GASTOS IMPUTADOS DIRECTAMENTE EN EL...,,3,-
3,TOTAL INGRESOS Y GASTOS IMPUTADOS DIRECTAMENTE...,,3,-
4,TRANSFERENCIASALACUENTADE PÉRDIDAS Y GANANCIAS,,-,-
5,TOTAL TRANSFERENCIAS A LA CUENTA DE PÉRDIDAS Y...,,-,-
6,TOTAL INGRESOS Y GASTOS RECONOCIDOS,,107.475,151.362



✔ Total dataframes extraídos: 5


---

## Enfoque 2 (adoptado) — PDF convertido a imágenes página a página

La solución que funciona: convertir cada página del PDF a una imagen PNG y aplicar el mismo pipeline que en `imagen_pipeline.ipynb`. Al tratar cada página como imagen independiente se resuelven los dos problemas del enfoque anterior:
- Las tablas ya no se cortan entre páginas: cada imagen contiene solo el fragmento del balance que cabe en una página.
- La densidad de filas por imagen es mucho menor que en el caso de imagen única, así que Docling no necesita recortes adicionales.


## Celda 2 — Conversión de páginas a imágenes y extracción con Docling

Lee las imágenes PNG de las páginas del PDF desde la carpeta de inputs (previamente convertidas con `pdf2image` o cualquier otra herramienta) y pasa cada una directamente a `DocumentConverter` de Docling sin recortes adicionales.

Se detectan **6 tablas** distribuidas entre las 3 páginas:
- **Página 1:** balance activo (activo no corriente + activo corriente).
- **Página 2:** balance pasivo y patrimonio neto + inicio de la PyG.
- **Página 3:** continuación de la PyG, resultado financiero y estado de cambios en el patrimonio neto.

Los DataFrames de todas las páginas se acumulan en `dfs_trozos`, con el mismo formato que en el pipeline de imagen.


In [5]:
from PIL import Image
import pandas as pd
import os
from docling.document_converter import DocumentConverter
from IPython.display import display

# ── CONFIGURACIÓN ─────────────────────────────────────────────────────────────
INPUTS = r"C:\Users\perdi\Desktop\tfm\TFM PYTHON\carpetaimagenes"
# ──────────────────────────────────────────────────────────────────────────────

imagenes = sorted([
    os.path.join(INPUTS, f)
    for f in os.listdir(INPUTS)
    if f.lower().endswith(".png")
])
print(f"Imágenes encontradas: {len(imagenes)}")

converter = DocumentConverter()

dfs_trozos = []
for ruta_imagen in imagenes:
    print(f"\n{'='*50}\n📄 {os.path.basename(ruta_imagen)}")
    result = converter.convert(ruta_imagen)
    print(f"  Tablas detectadas: {len(result.document.tables)}")
    for i, tabla in enumerate(result.document.tables, 1):
        print(f"\n── Tabla {i} ──")
        df_trozo = tabla.export_to_dataframe(doc=result.document)
        df_trozo = pd.DataFrame(
            [df_trozo.columns.tolist()] + df_trozo.values.tolist()
        )
        df_trozo.columns = [f"C{j}" for j in range(df_trozo.shape[1])]
        display(df_trozo)
        dfs_trozos.append(df_trozo)

print(f"\n✔ Total dataframes extraídos: {len(dfs_trozos)}")

[INFO] 2026-05-06 10:29:53,680 [RapidOCR] base.py:22: Using engine_name: onnxruntime
[INFO] 2026-05-06 10:29:53,686 [RapidOCR] download_file.py:60: File exists and is valid: C:\Users\perdi\Desktop\tfm\TFM PYTHON\entornotfm\Lib\site-packages\rapidocr\models\ch_PP-OCRv4_det_mobile.onnx
[INFO] 2026-05-06 10:29:53,687 [RapidOCR] main.py:57: Using C:\Users\perdi\Desktop\tfm\TFM PYTHON\entornotfm\Lib\site-packages\rapidocr\models\ch_PP-OCRv4_det_mobile.onnx
[INFO] 2026-05-06 10:29:53,791 [RapidOCR] base.py:22: Using engine_name: onnxruntime
[INFO] 2026-05-06 10:29:53,793 [RapidOCR] download_file.py:60: File exists and is valid: C:\Users\perdi\Desktop\tfm\TFM PYTHON\entornotfm\Lib\site-packages\rapidocr\models\ch_ppocr_mobile_v2.0_cls_mobile.onnx
[INFO] 2026-05-06 10:29:53,794 [RapidOCR] main.py:57: Using C:\Users\perdi\Desktop\tfm\TFM PYTHON\entornotfm\Lib\site-packages\rapidocr\models\ch_ppocr_mobile_v2.0_cls_mobile.onnx


Imágenes encontradas: 3

📄 pagina_001.png


[INFO] 2026-05-06 10:29:53,850 [RapidOCR] base.py:22: Using engine_name: onnxruntime
[INFO] 2026-05-06 10:29:53,860 [RapidOCR] download_file.py:60: File exists and is valid: C:\Users\perdi\Desktop\tfm\TFM PYTHON\entornotfm\Lib\site-packages\rapidocr\models\ch_PP-OCRv4_rec_mobile.onnx
[INFO] 2026-05-06 10:29:53,861 [RapidOCR] main.py:57: Using C:\Users\perdi\Desktop\tfm\TFM PYTHON\entornotfm\Lib\site-packages\rapidocr\models\ch_PP-OCRv4_rec_mobile.onnx
Loading weights: 100%|██████████| 770/770 [00:00<00:00, 9757.91it/s]


  Tablas detectadas: 1

── Tabla 1 ──


,C0,C1,C2,C3
0,ACTIVO,Notas,Periodo.2024,Periodo.2023
1,InmovilizadoIntangible,5,8.846,8.862
2,Aplicacionesinformaticas,,8.769,8.862
3,Anticiposy activos en curso,,77,0
4,InmovilizadoMaterial,,924,1.246
5,Instalacionestecnicasy otroinmovilizado material,,923,1.246
6,Anticiposyactivosen curso,,,0
7,Inversionesfinancierasenempresasdel grupo,,560.000,548.028
8,Instrumentosdepatrimonio,6,560.000,547.236
9,Creditosaempresas,7,0,792



📄 pagina_002.png
  Tablas detectadas: 2

── Tabla 1 ──


,C0,C1,C2,C3
0,Capital,9.1,32.550,32.550
1,Capital escriturado,,32.550,32.550
2,Prima de emision,9.2,12,12
3,Reservas,6,500.160,523.464
4,,9.5,-35.045,-21.671
5,...Resultado.del.ejercicio.,3,107.472,.151.362.
6,Dividendo a cuenta,3.1,-26.844,64.563..
7,...Otros.instrumentos.e.patrimonio.,9.6,3.696..,2.358.
8,"...Subvenciones,.donaciones.y.legados.",,3..,
9,..ATRIMONIO..NETO.,,582.004..,623.512



── Tabla 2 ──


,C0,C1,C2,C3
0,(Miles de euros).,Notas,Periodo.2.024,Periodo.2.023
1,OPERACIONES CONTINUADAS,,,
2,Importe neto de la cifra de negocios,13.1,153.786,191.750
3,patrimonio Ingresos de participaciones en inst...,,122.363,161.853
4,Prestacion de servicios y otros ingresos,,31.423,29.897
5,Aprovisionamientos,,-440,-184
6,Consumo de mercaderias,13.2,-440,-184
7,Otros ingresos de explotacion,,22,31
8,Subvenciones de explotacion incorporadas al re...,,22,31
9,Gastos de personal,13.3,-23.901,-21.529



📄 pagina_003.png
  Tablas detectadas: 3

── Tabla 1 ──


,C0,C1,C2,C3
0,Subvenciones de explotacion incorporadas al re...,,22,31
1,Gastos de personal,13.3,-23.901,-21.529
2,"Sueldos, salarios y asimilados",,-20.332,-17.813
3,Cargas sociales,,-3.569,-3.716
4,Otros gastos de explotacion,,-12.742,-11.305
5,Servicios exteriores,13.4,-12.737,-11.297
6,,,-5,8-
7,Amortizacion de inmovilizado,5,-3.202,-3.019
8,Deterioroyresultado porenajenaciones de inmovi...,,-4,5
9,Resultados por enajenaciones y otras,,-4,5



── Tabla 2 ──


,C0,C1,C2,C3
0,RESULTADO FINANCIERO,,-2.047,-52
1,RESULTADOANTES DEIMPUESTOS,,111.472,155.697
2,Gasto por impuesto sobre las ganancias,12,-4.000,-4.335
3,OPERACIONESCONTINUADAS RESULTADODELEJERCICIOPR...,,107.472..,151.362
4,RESULTADODELEJERCICIO,,107.472,151.362



── Tabla 3 ──


,C0,C1,C2,C3
0,(Miles de euros),Notas,Periodo.2024,Periodo.2023
1,RESULTADODE LA CUENTA DEPERDIDASY GANANCIAS,,107.472,151.362
2,INGRESOS Y GASTOSIMPUTADOS DIRECTAMENTE EN EL ...,,3,
3,TOTAL INGRESOS Y GASTOS IMPUTADOS DIRECTAMENTE...,,3,
4,TRANSFERENCIAS A LA CUENTA DE PERDIDAS Y GANAN...,,,
5,TOTALTRANSFERENCIASA LA CUENTA DEPERDIDASY GAN...,,,
6,TOTALINGRESOSYGASTOSRECONOCIDOS,,107.475,151.362



✔ Total dataframes extraídos: 6


## Celda 3 — Inspección de los DataFrames extraídos

Muestra todas las tablas para verificar la calidad de la extracción. Se aprecia el **ruido OCR** característico de PDFs con tipografías compactas:

- Palabras concatenadas: `"InmovilizadoIntangible"`, `"ACTIVOSCORRIENTES"`, `"Deudorescomercialesyotrascuentasacobrar"`.
- Puntos intercalados: `"...Resultado.del.ejercicio."`, `"..ATRIMONIO..NETO."`.
- Letras iniciales perdidas: `"..PASIVO.NO.CORRIENTE"`.

Este nivel de ruido es manejable para el LLM de selección de partidas, que está instruido explícitamente para tolerarlo y reconocer las partidas aunque vengan con grafías no estándar.


In [6]:
for i, df in enumerate(dfs_trozos, 1):
    print(f"\n── Tabla {i} ──")
    display(df)


── Tabla 1 ──


,C0,C1,C2,C3
0,ACTIVO,Notas,Periodo.2024,Periodo.2023
1,InmovilizadoIntangible,5,8.846,8.862
2,Aplicacionesinformaticas,,8.769,8.862
3,Anticiposy activos en curso,,77,0
4,InmovilizadoMaterial,,924,1.246
5,Instalacionestecnicasy otroinmovilizado material,,923,1.246
6,Anticiposyactivosen curso,,,0
7,Inversionesfinancierasenempresasdel grupo,,560.000,548.028
8,Instrumentosdepatrimonio,6,560.000,547.236
9,Creditosaempresas,7,0,792



── Tabla 2 ──


,C0,C1,C2,C3
0,Capital,9.1,32.550,32.550
1,Capital escriturado,,32.550,32.550
2,Prima de emision,9.2,12,12
3,Reservas,6,500.160,523.464
4,,9.5,-35.045,-21.671
5,...Resultado.del.ejercicio.,3,107.472,.151.362.
6,Dividendo a cuenta,3.1,-26.844,64.563..
7,...Otros.instrumentos.e.patrimonio.,9.6,3.696..,2.358.
8,"...Subvenciones,.donaciones.y.legados.",,3..,
9,..ATRIMONIO..NETO.,,582.004..,623.512



── Tabla 3 ──


,C0,C1,C2,C3
0,(Miles de euros).,Notas,Periodo.2.024,Periodo.2.023
1,OPERACIONES CONTINUADAS,,,
2,Importe neto de la cifra de negocios,13.1,153.786,191.750
3,patrimonio Ingresos de participaciones en inst...,,122.363,161.853
4,Prestacion de servicios y otros ingresos,,31.423,29.897
5,Aprovisionamientos,,-440,-184
6,Consumo de mercaderias,13.2,-440,-184
7,Otros ingresos de explotacion,,22,31
8,Subvenciones de explotacion incorporadas al re...,,22,31
9,Gastos de personal,13.3,-23.901,-21.529



── Tabla 4 ──


,C0,C1,C2,C3
0,Subvenciones de explotacion incorporadas al re...,,22,31
1,Gastos de personal,13.3,-23.901,-21.529
2,"Sueldos, salarios y asimilados",,-20.332,-17.813
3,Cargas sociales,,-3.569,-3.716
4,Otros gastos de explotacion,,-12.742,-11.305
5,Servicios exteriores,13.4,-12.737,-11.297
6,,,-5,8-
7,Amortizacion de inmovilizado,5,-3.202,-3.019
8,Deterioroyresultado porenajenaciones de inmovi...,,-4,5
9,Resultados por enajenaciones y otras,,-4,5



── Tabla 5 ──


,C0,C1,C2,C3
0,RESULTADO FINANCIERO,,-2.047,-52
1,RESULTADOANTES DEIMPUESTOS,,111.472,155.697
2,Gasto por impuesto sobre las ganancias,12,-4.000,-4.335
3,OPERACIONESCONTINUADAS RESULTADODELEJERCICIOPR...,,107.472..,151.362
4,RESULTADODELEJERCICIO,,107.472,151.362



── Tabla 6 ──


,C0,C1,C2,C3
0,(Miles de euros),Notas,Periodo.2024,Periodo.2023
1,RESULTADODE LA CUENTA DEPERDIDASY GANANCIAS,,107.472,151.362
2,INGRESOS Y GASTOSIMPUTADOS DIRECTAMENTE EN EL ...,,3,
3,TOTAL INGRESOS Y GASTOS IMPUTADOS DIRECTAMENTE...,,3,
4,TRANSFERENCIAS A LA CUENTA DE PERDIDAS Y GANAN...,,,
5,TOTALTRANSFERENCIASA LA CUENTA DEPERDIDASY GAN...,,,
6,TOTALINGRESOSYGASTOSRECONOCIDOS,,107.475,151.362


## Celda 4 — Detección de layout por tabla

Para cada tabla extraída por Docling se serializa su contenido y se envía al LLM. El prompt:
- Pide que identifique la columna de partidas y las columnas de valores numéricos.
- Pide que detecte las fechas directamente en la tabla (a diferencia del pipeline de imagen, aquí no hay una celda separada de extracción de fechas con visión porque la cabecera está en cada tabla del PDF).
- Maneja la peculiaridad de este PDF: las columnas de fecha aparecen como `"Periodo.2024"` y `"Periodo.2023"` en lugar de fechas explícitas.

Para tablas de páginas siguientes donde no se detectan fechas propias, el código hereda las fechas de la primera tabla que sí las tenga.


In [33]:
import os
from dotenv import load_dotenv
import openai
import json

load_dotenv()

GROQ_API_KEY  = os.getenv("GROQ_API_KEY_2", "")
GROQ_BASE_URL = "https://api.groq.com/openai/v1"
MODELO        = "openai/gpt-oss-120b"

client = openai.OpenAI(api_key=GROQ_API_KEY, base_url=GROQ_BASE_URL)

layouts = {}

for i, df in enumerate(dfs_trozos):
    print(f"\n{'='*50}")
    print(f"── Tabla {i+1}/{len(dfs_trozos)} ──")

    lineas = []
    for row_idx, row in df.head(15).iterrows():
        for col_name, val in row.items():
            if pd.notna(val) and str(val).strip() not in ("", "nan"):
                lineas.append(f"fila={row_idx}, col={col_name}: {val}")

    mapa = "\n".join(lineas)

    prompt = f"""Tienes un fragmento de un balance financiero extraído con OCR, mapeado celda a celda.
Cada línea indica exactamente en qué fila y columna está cada valor no vacío.

{mapa}

Analiza la estructura y responde SOLO en JSON sin markdown ni backticks:

{{
  "fechas_detectadas": {{"nombre_columna": "fecha exacta encontrada, ej: 31/12/2024"}},
  "unidades": "mil EUR / EUR / desconocido",
  "columna_partidas": "columna exacta con los nombres de las partidas",
  "columnas_valores": ["columnas con valores numéricos"],
  "razonamiento": "breve explicación",
  "ejemplo": {{
    "partida": "nombre de partida de ejemplo",
    "valores": {{"nombre_columna": valor_numerico}}
  }}
}}

Instrucciones clave:
- En fechas_detectadas pon la fecha REAL que hayas encontrado en el mapa (ej: "C3": "31/12/2024"), no el texto "fecha".
- Las fechas (31/12/2024, 31/12/2023...) y unidades (mil EUR) aparecen en unas columnas, pero los valores numéricos del balance pueden estar en columnas DISTINTAS. No asumas que coinciden.
- Los valores numéricos son importes grandes del balance (ej: 12497590, 11416223).
- La columna de partidas contiene descripciones contables: 'Activo', 'Inmovilizado', 'Existencias', etc.
- Puede haber columnas intermedias vacías o con metadatos, ignóralas.
- Fíjate bien en qué columna aparecen los números grandes para identificar columnas_valores.
- Fíjate bien en qué columna aparecen las fechas para identificar fechas_detectadas.
- Las columnas de fecha y de valor pueden estar desplazadas una respecto a la otra.
- En este caso las fechas pueden aparecer como "Periodo.2024" o "Periodo.2023", tratalas como 31/12/2024 y 31/12/2023.
"""

    response = client.chat.completions.create(
        model=MODELO,
        messages=[{"role": "user", "content": prompt}],
        temperature=0
    )
    print(f"Input tokens:  {response.usage.prompt_tokens}")
    print(f"Output tokens: {response.usage.completion_tokens}")
    print(f"Total tokens:  {response.usage.total_tokens}")


    raw = response.choices[0].message.content.strip().strip("```json").strip("```").strip()

    try:
        layout = json.loads(raw)
        layouts[i] = layout
        print(f"Fechas:           {layout['fechas_detectadas']}")
        print(f"Unidades:         {layout['unidades']}")
        print(f"Columna partidas: {layout['columna_partidas']}")
        print(f"Columnas valores: {layout['columnas_valores']}")
        print(f"Ejemplo:          {layout['ejemplo']}")
    except json.JSONDecodeError:
        print(f"⚠️ No devolvió JSON limpio:")
        print(raw)


── Tabla 1/6 ──
Input tokens:  1159
Output tokens: 737
Total tokens:  1896
Fechas:           {'C2': '31/12/2024', 'C3': '31/12/2023'}
Unidades:         desconocido
Columna partidas: C0
Columnas valores: ['C2', 'C3']
Ejemplo:          {'partida': 'InmovilizadoIntangible', 'valores': {'C2': 8846, 'C3': 8862}}

── Tabla 2/6 ──
Input tokens:  1191
Output tokens: 722
Total tokens:  1913
Fechas:           {}
Unidades:         desconocido
Columna partidas: C0
Columnas valores: ['C2', 'C3']
Ejemplo:          {'partida': 'Capital', 'valores': {'C2': 32.55, 'C3': 32.55}}

── Tabla 3/6 ──
Input tokens:  1141
Output tokens: 524
Total tokens:  1665
Fechas:           {'C2': '31/12/2024', 'C3': '31/12/2023'}
Unidades:         mil EUR
Columna partidas: C0
Columnas valores: ['C2', 'C3']
Ejemplo:          {'partida': 'Importe neto de la cifra de negocios', 'valores': {'C2': 153.786, 'C3': 191.75}}

── Tabla 4/6 ──
Input tokens:  1138
Output tokens: 816
Total tokens:  1954
Fechas:           {}
Unidades:

## Celda 5 — Construcción del DataFrame de conceptos

Con el layout detectado se construye el DataFrame largo `{partida, fecha, valor, tabla}`. Se aplica limpieza de valores numéricos específica para el ruido de este PDF:

- `val.replace(":", ".")` — corrige confusiones OCR entre `:` y `.` en decimales (`58:190` → `58.190`).
- `val.strip(".")` — elimina puntos al inicio o final que son artefactos de líneas de tabla (`".151.362."` → `151.362`).
- `float(val.replace(".", "").replace(",", "."))` — convierte el formato español de miles (punto como separador de miles, coma como decimal) a float de Python.


In [26]:
registros = []

for i, df in enumerate(dfs_trozos):
    if i not in layouts:
        print(f"⚠️ Tabla {i+1} sin layout, saltando")
        continue

    layout       = layouts[i]
    col_partidas = layout['columna_partidas']
    cols_valores = layout['columnas_valores']
    fechas       = layout['fechas_detectadas']

    if not fechas:
        for j in range(len(dfs_trozos)):
            if j in layouts and layouts[j]['fechas_detectadas']:
                fechas = layouts[j]['fechas_detectadas']
                break

    col_a_fecha = {cv: cf for cv, cf in zip(cols_valores, fechas.values())}

    for _, row in df.iterrows():
        partida = str(row.get(col_partidas, "")).strip()
        if not partida or partida in ("nan", ""):
            continue
        for col_val, fecha in col_a_fecha.items():
            if col_val not in df.columns:
                continue
            val = str(row.get(col_val, "")).strip()
            val = val.replace(":", ".")   # 58:190 → 58.190
            val = val.strip(".")          # .151.362. → 151.362
            val = val.strip()
            try:
                val = float(val.replace(".", "").replace(",", "."))
            except ValueError:
                val = None
            registros.append({"partida": partida, "fecha": fecha, "valor": val, "tabla": i+1})

df_resultado = pd.DataFrame(registros)
pd.set_option('display.max_rows', None)
display(df_resultado)
pd.reset_option('display.max_rows')

,partida,fecha,valor,tabla
0,ACTIVO,31/12/2024,NaN,1
1,ACTIVO,31/12/2023,NaN,1
2,InmovilizadoIntangible,31/12/2024,8846.0,1
3,InmovilizadoIntangible,31/12/2023,8862.0,1
4,Aplicacionesinformaticas,31/12/2024,8769.0,1
5,Aplicacionesinformaticas,31/12/2023,8862.0,1
6,Anticiposy activos en curso,31/12/2024,77.0,1
7,Anticiposy activos en curso,31/12/2023,0.0,1
8,InmovilizadoMaterial,31/12/2024,924.0,1
9,InmovilizadoMaterial,31/12/2023,1246.0,1


---

## Selección de partidas — LLM con índice posicional

Con el DataFrame construido, el siguiente paso es identificar cuál de todas las partidas extraídas corresponde a cada uno de los 22 alias que necesitamos para calcular los ratios.

**Enfoque:** se construye un índice `tabla=N, fila=M: texto_partida` con todas las partidas en orden y se pasa al LLM en una sola llamada. El modelo razona por contexto jerárquico (qué sección del balance rodea a cada partida) para resolver los casos ambiguos sin necesidad de ningún preprocessing adicional.

Por qué este enfoque y no fuzzy matching: el fuzzy matching falla sistemáticamente con partidas de nombre idéntico en secciones distintas del balance (`"Deudas con entidades de crédito"` aparece igual en LP y CP) y con el ruido OCR de este PDF, donde las partidas pueden estar tan distorsionadas que la similitud de caracteres colapsa. Pasar el índice directamente al LLM resuelve ambos problemas de una vez y es más simple.


## Celda 6 — Selección de las 22 partidas mediante LLM con índice posicional

Construye el índice completo de partidas y lo envía al LLM con instrucciones adaptadas al ruido OCR de este PDF concreto:

- Menciona variantes OCR esperadas para las partidas más conflictivas (`"..ATRIMONIO..NETO."`, `"TOTALACTIVO"`, `"ACTIVOSCORRIENTES"`).
- Para las deudas bancarias LP/CP usa el razonamiento por anclas de sección (qué cabecera de bloque aparece más cerca en número de fila) para distinguir entre dos partidas con nombre idéntico.
- Incluye un ejemplo de output con las grafías con ruido real que aparecen en este documento.

El output es un JSON `{alias: {tabla, fila, partida}}` con los 22 mapeos. El coste de esta llamada es de ~3.000-5.000 tokens, comparable al enfoque de imagen.


In [ ]:
# Construir lista ordenada de partidas de TODAS las tablas
lineas_partidas = []
for i, df in enumerate(dfs_trozos):
    if i not in layouts:
        continue
    col_partidas = layouts[i]['columna_partidas']
    for row_idx, row in df.iterrows():
        partida = str(row.get(col_partidas, "")).strip()
        if partida and partida not in ("nan", ""):
            lineas_partidas.append(f"tabla={i+1}, fila={row_idx}: {partida}")
partidas_formateadas = "\n".join(lineas_partidas)

prompt_mapeo = f"""Eres un experto en contabilidad española (PGC 2007).
Tienes la lista completa de partidas de un balance financiero extraído con OCR, 
en orden exacto tal como aparecen, con su tabla y fila.

PARTIDAS DISPONIBLES:
{partidas_formateadas}

INSTRUCCIONES GENERALES:
- El OCR puede haber introducido ruido: puntos, guiones, letras pegadas, apóstrofes.
- Devuelve EXACTAMENTE el texto de la partida tal como aparece en la lista.
- Si una partida no existe en este balance devuelve null.

INSTRUCCIONES ESPECIALES:
- 'patrimonio_neto' y 'fondos_propios': puede aparecer como '..ATRIMONIO..NETO.' sin la P, eligelo igualmente. Usa la misma partida para ambos.
- 'pasivo_no_corriente': puede aparecer como '..PASIVO.NO.CORRIENTE'.
- 'pasivo_corriente': puede aparecer como '..PASIVO...CORRIENTE..'.
- 'total_activo': puede aparecer como 'TOTALACTIVO'.
- 'activo_corriente': puede aparecer como 'ACTIVOSCORRIENTES'.
- 'resultado_ejercicio': puede aparecer como '...Resultado.del.ejercicio.' o 'RESULTADODELEJERCICIO'.
- 'efectivo': puede aparecer como 'Efectivoy equivalentes' o 'Tesoreria', elige la primera.
- 'deuda_credito_lp' y 'deuda_credito_cp': busca la partida "Deudas con entidades de credito"
  (puede aparecer con ruido OCR como "Deudasconentidadesde'credito").
  Una vez localizada, mira las filas inmediatamente anteriores y posteriores y determina
  si esta ENTRE o MAS CERCA de alguna de estas anclas de CP:
  "Deudas a corto plazo", "corto plazo", "PASIVO CORRIENTE", "pasivo corriente"
  o de estas anclas de LP:
  "Deudas a largo plazo", "largo plazo", "PASIVO NO CORRIENTE", "pasivo no corriente".
  La ancla que aparezca mas cerca en numero de filas determina si es LP o CP.
  Si es CP → asigna a deuda_credito_cp y pon null en deuda_credito_lp.
  Si es LP → asigna a deuda_credito_lp y pon null en deuda_credito_cp.
  IMPORTANTE: "Deudas a corto plazo" y "Deudas a largo plazo" tambien cuentan como anclas,
  no solo "PASIVO CORRIENTE" y "PASIVO NO CORRIENTE".

Ejemplo de output:
{{
  "patrimonio_neto":            {{"tabla": 2, "fila": 9,  "partida": "..ATRIMONIO..NETO."}},
  "fondos_propios":             {{"tabla": 2, "fila": 9,  "partida": "..ATRIMONIO..NETO."}},
  "pasivo_no_corriente":        {{"tabla": 2, "fila": 17, "partida": "..PASIVO.NO.CORRIENTE"}},
  "deudas_lp":                  {{"tabla": 2, "fila": 12, "partida": "Deudas-a largo-plazo"}},
  "deudas_cp":                  {{"tabla": 2, "fila": 18, "partida": "Deudas a corto plazo"}},
  "pasivo_corriente":           {{"tabla": 2, "fila": 30, "partida": "..PASIVO...CORRIENTE.."}},
  "total_activo":               {{"tabla": 1, "fila": 33, "partida": "TOTALACTIVO"}},
  "activo_corriente":           {{"tabla": 1, "fila": 32, "partida": "ACTIVOSCORRIENTES"}},
  "existencias":                null,
  "deudores_comerciales":       {{"tabla": 1, "fila": 16, "partida": "Deudorescomercialesyotrascuentasacobrar"}},
  "efectivo":                   {{"tabla": 1, "fila": 30, "partida": "Efectivoy equivalentes"}},
  "inmovilizado_material":      {{"tabla": 1, "fila": 4,  "partida": "InmovilizadoMaterial"}},
  "acreedores_comerciales":     {{"tabla": 2, "fila": 23, "partida": "Acreedorescomercialesyotrascuentasapagar"}},
  "cifra_negocios":             {{"tabla": 3, "fila": 2,  "partida": "Importe neto de la cifra de negocios"}},
  "otros_ingresos_explotacion": {{"tabla": 3, "fila": 7,  "partida": "Otros ingresos de explotacion"}},
  "resultado_explotacion":      {{"tabla": 3, "fila": 18, "partida": "RESULTADO DE EXPLOTACION"}},
  "amortizacion":               {{"tabla": 3, "fila": 15, "partida": "Amortizacion deinmovilizado"}},
  "gastos_financieros":         {{"tabla": 3, "fila": 19, "partida": "Gastos financieros"}},
  "resultado_antes_impuestos":  {{"tabla": 5, "fila": 1,  "partida": "RESULTADOANTES DEIMPUESTOS"}},
  "resultado_ejercicio":        {{"tabla": 2, "fila": 5,  "partida": "...Resultado.del.ejercicio."}},
  "deuda_credito_lp":           null,
  "deuda_credito_cp":           {{"tabla": 2, "fila": 19, "partida": "Deudasconentidadesde'credito"}}
}}

Ahora analiza las partidas disponibles de este balance y devuelve el JSON con los valores reales. SOLO JSON sin markdown.
"""

response = client.chat.completions.create(
    model=MODELO,
    messages=[{"role": "user", "content": prompt_mapeo}],
    temperature=0
)
print(f"Input tokens:  {response.usage.prompt_tokens}")
print(f"Output tokens: {response.usage.completion_tokens}")
print(f"Total tokens:  {response.usage.total_tokens}")

raw = response.choices[0].message.content.strip().strip("```json").strip("```").strip()
print(raw)

try:
    mapeo_llm = json.loads(raw)
    print("\n── Mapeo LLM ──")
    for alias, info in mapeo_llm.items():
        if info:
            print(f"  {alias}: tabla={info['tabla']}, fila={info['fila']} → {info['partida']}")
        else:
            print(f"  {alias}: null")
except json.JSONDecodeError:
    print("⚠️ No devolvió JSON limpio")
    print(raw)

Input tokens:  3184
Output tokens: 1916
Total tokens:  5100
{
  "patrimonio_neto":            {"tabla": 2, "fila": 9,  "partida": "..ATRIMONIO..NETO."},
  "fondos_propios":             {"tabla": 2, "fila": 9,  "partida": "..ATRIMONIO..NETO."},
  "pasivo_no_corriente":        {"tabla": 2, "fila": 17, "partida": "..PASIVO.NO.CORRIENTE"},
  "deudas_lp":                  {"tabla": 2, "fila": 12, "partida": "Deudas-a largo-plazo"},
  "deudas_cp":                  {"tabla": 2, "fila": 18, "partida": "Deudas a corto plazo"},
  "pasivo_corriente":           {"tabla": 2, "fila": 30, "partida": "..PASIVO...CORRIENTE.."},
  "total_activo":               {"tabla": 1, "fila": 33, "partida": "TOTALACTIVO"},
  "activo_corriente":           {"tabla": 1, "fila": 32, "partida": "ACTIVOSCORRIENTES"},
  "existencias":                null,
  "deudores_comerciales":       {"tabla": 1, "fila": 16, "partida": "Deudorescomercialesyotrascuentasacobrar"},
  "efectivo":                   {"tabla": 1, "fila": 30, 

## Celda 7 — Extracción de valores por período

Con el mapeo del LLM se recuperan los valores cruzando `(tabla_idx, fila_idx)` con los DataFrames. Se aplica la misma limpieza de valores numéricos que en la celda 5 (puntos/comas como separadores de miles/decimales, artefactos OCR).

El resultado es `valores_por_fecha`, con la misma estructura que en los pipelines Excel e imagen.


In [30]:
valores_por_fecha = {}

for alias, info in mapeo_llm.items():
    if info is None:
        continue

    tabla_idx    = info['tabla'] - 1  # 0-indexed
    fila_idx     = info['fila']

    if tabla_idx >= len(dfs_trozos) or tabla_idx not in layouts:
        print(f"⚠️ {alias}: tabla {tabla_idx+1} no disponible")
        continue

    df           = dfs_trozos[tabla_idx]
    layout       = layouts[tabla_idx]
    cols_valores = layout['columnas_valores']
    fechas       = layout['fechas_detectadas']

    if not fechas:
        for j in range(len(dfs_trozos)):
            if j in layouts and layouts[j]['fechas_detectadas']:
                fechas = layouts[j]['fechas_detectadas']
                break

    col_a_fecha = {cv: cf for cv, cf in zip(cols_valores, fechas.values())}

    if fila_idx not in df.index:
        print(f"⚠️ {alias}: fila {fila_idx} no existe en tabla {tabla_idx+1}")
        continue

    row = df.loc[fila_idx]
    for col_val, fecha in col_a_fecha.items():
        if col_val not in df.columns:
            continue
        val = str(row.get(col_val, "")).strip()
        val = val.replace(":", ".").strip(".").strip()
        try:
            val = float(val.replace(".", "").replace(",", "."))
        except ValueError:
            val = None

        if fecha not in valores_por_fecha:
            valores_por_fecha[fecha] = {}
        if alias not in valores_por_fecha[fecha]:
            valores_por_fecha[fecha][alias] = val

print(pd.DataFrame(valores_por_fecha).T)

            patrimonio_neto  fondos_propios  pasivo_no_corriente  deudas_lp  \
31/12/2024         582004.0        582004.0                 13.0       42.0   
31/12/2023         623512.0        623512.0               2104.0       20.0   

            deudas_cp  pasivo_corriente  total_activo  activo_corriente  \
31/12/2024    60492.0           78128.0      660145.0           85530.0   
31/12/2023    17601.0           32118.0      657734.0           94203.0   

            deudores_comerciales  efectivo  inmovilizado_material  \
31/12/2024               37773.0     821.0                  924.0   
31/12/2023               33787.0     559.0                 1246.0   

            acreedores_comerciales  cifra_negocios  \
31/12/2024                 15570.0        153786.0   
31/12/2023                 14483.0        191750.0   

            otros_ingresos_explotacion  resultado_explotacion  amortizacion  \
31/12/2024                        22.0               113519.0       -3202.0   
31/12/2

## Celda 8 — Cálculo de 22 ratios financieros — resultado final

Cálculo de los 22 ratios sobre `valores_por_fecha`. El código es idéntico al de los otros dos pipelines con una única diferencia:

```python
amort = abs(v.get('amortizacion') or 0)
```

`abs()` es necesario porque en la PyG de Viscofan la amortización aparece con signo negativo (es un gasto contable). Sin él, `EBITDA = EBIT - (-3.202) = EBIT + 3.202`, inflando el EBITDA en el doble de la amortización y produciendo un ICR y un margen EBITDA incorrectamente elevados.

**Validación contra datos reales de Viscofan 2024:** 20 de los 22 ratios son exactos. Las dos diferencias (ICR y margen EBITDA) se deben a un error residual de OCR en la amortización sobre este PDF concreto — la partida se extrae con un valor ligeramente distinto al real por las líneas superpuestas de la tabla en el documento fuente. Esto ilustra la limitación documentada en la memoria sobre la sensibilidad al ruido en PDFs de baja calidad de exportación.


In [31]:
def safe_div(a, b):
    if a is None or b is None or b == 0: return None
    return round(a / b, 6)

def pct(a, b):   r = safe_div(a, b); return round(r * 100, 2) if r is not None else None
def ratio(a, b): r = safe_div(a, b); return round(r, 2)       if r is not None else None
def dias(a, b):  r = safe_div(a, b); return round(r * 365, 1) if r is not None else None

ratios_calculados = {}

for fecha, v in valores_por_fecha.items():
    cn    = v.get('cifra_negocios')
    oi    = v.get('otros_ingresos_explotacion') or 0
    ebit  = v.get('resultado_explotacion')
    amort = abs(v.get('amortizacion') or 0)  # abs() porque en PyG viene con signo negativo
    gf    = v.get('gastos_financieros')
    res   = v.get('resultado_ejercicio')
    act   = v.get('total_activo')
    actc  = v.get('activo_corriente')
    exst  = v.get('existencias') or 0
    deud  = v.get('deudores_comerciales')
    efec  = v.get('efectivo') or 0
    fp    = v.get('fondos_propios')
    pasc  = v.get('pasivo_corriente')
    acr   = v.get('acreedores_comerciales')
    delp  = v.get('deuda_credito_lp') or 0
    decp  = v.get('deuda_credito_cp') or 0
    pnc   = v.get('pasivo_no_corriente')

    ebitda  = (ebit - amort) if ebit  is not None else None
    gf_abs  = abs(gf)        if gf    is not None else None
    pas_tot = (pnc + pasc)   if (pnc  is not None and pasc is not None) else None
    fm      = (actc - pasc)  if (actc is not None and pasc is not None) else None
    deuda_f = delp + decp
    dfn     = deuda_f - efec

    pmc   = dias(deud, cn)
    pmp   = dias(acr,  cn)
    stock = dias(exst, cn)
    ccc   = round((pmc or 0) + (stock or 0) - (pmp or 0), 1) if all(x is not None for x in [pmc, pmp, stock]) else None

    ratios_calculados[fecha] = {
        "SOL01_solvencia":              ratio(act, pas_tot),
        "SOL02_autonomia_financiera":   pct(fp, act),
        "SOL03_endeudamiento":          ratio(pas_tot, fp),
        "SOL04_dfn_sobre_fp":           pct(dfn, fp),
        "COV01_dfn_sobre_ebitda":       ratio(dfn, ebitda),
        "COV02_icr":                    ratio(ebitda, gf_abs),
        "COV03_carga_financiera_cn":    pct(gf_abs, cn),
        "LIQ01_ratio_corriente":        ratio(actc, pasc),
        "LIQ02_ratio_acido":            ratio((actc - exst) if actc is not None else None, pasc),
        "LIQ03_fm_sobre_cn":            pct(fm, cn),
        "LIQ04_tesoreria_sobre_activo": pct(efec, act),
        "REN01_margen_ebitda":          pct(ebitda, cn),
        "REN02_margen_ebit":            pct(ebit, cn),
        "REN03_margen_neto":            pct(res, cn),
        "REN04_roa":                    pct(ebit, act),
        "REN05_roe":                    pct(res, fp),
        "APL01_deuda_fin_sobre_activo": pct(deuda_f, act),
        "APL02_pasivo_sobre_activo":    pct(pas_tot, act),
        "EFI01_pmc":   pmc,
        "EFI02_pmp":   pmp,
        "EFI03_stock": stock,
        "EFI04_ccc":   ccc,
    }

df_ratios = pd.DataFrame(ratios_calculados).T
df_ratios = pd.DataFrame(ratios_calculados).T

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.width', None)
pd.set_option('display.float_format', '{:.2f}'.format)
display(df_ratios)
pd.reset_option('display.max_columns')
pd.reset_option('display.max_rows')
pd.reset_option('display.width')
pd.reset_option('display.float_format')

,SOL01_solvencia,SOL02_autonomia_financiera,SOL03_endeudamiento,SOL04_dfn_sobre_fp,COV01_dfn_sobre_ebitda,COV02_icr,COV03_carga_financiera_cn,LIQ01_ratio_corriente,LIQ02_ratio_acido,LIQ03_fm_sobre_cn,LIQ04_tesoreria_sobre_activo,REN01_margen_ebitda,REN02_margen_ebit,REN03_margen_neto,REN04_roa,REN05_roe,APL01_deuda_fin_sobre_activo,APL02_pasivo_sobre_activo,EFI01_pmc,EFI02_pmp,EFI03_stock,EFI04_ccc
31/12/2024,8.45,88.16,0.13,9.86,0.49,87.37,0.87,1.09,1.09,4.81,0.12,75.90,73.82,69.88,17.20,18.47,8.81,11.84,89.70,37.00,0.00,52.70
31/12/2023,19.22,94.80,0.05,2.28,0.09,465.60,0.18,2.93,2.93,32.38,0.08,82.80,81.23,78.94,23.68,24.28,2.25,5.20,64.30,27.60,0.00,36.70
